Do not Consume Extra Token in DicomJsonConverter

wsugarman opened the pull request:
Fixes #1251.

#### Checklist
- [x] The pull request branch is in sync with latest commit on the *fo-dicom/development* branch
- [x] No API documentation update needed
- [x] I have included unit tests
- [x] I have updated the change log
- [x] I am listed in the CONTRIBUTORS file

#### Changes proposed in this pull request:
- Fix `DicomJsonConverter` to no longer consume the end object JSON token. With this fix, the converter may be used when deserializing arrays automatically
- Mark the `DicomArrayJsonConverter` as `Obsolete`
- Throw a `JsonException` if the token is unexpected instead of returning `null`

mrbean-bremen reviewed (commented):
Thanks, that looks good to me! Can you also add a test for converting a json array (like you showed in the bug description)?

wsugarman commented:
Added some new tests. Now that `DicomJson.ConvertJsonToDicomArray` uses `DicomJsonConverter` directly, I've added cases for each test to also validate the `DicomArrayJsonConverter`.

wsugarman commented on FO-DICOM.Core/Serialization/Utf8JsonReaderExtensions.cs:
```diff
@@ -0,0 +1,24 @@
+{
+    internal static class Utf8JsonReaderExtensions
+    {
+        public static void Assume(this ref Utf8JsonReader reader, JsonTokenType tokenType)
```
I've made these extension methods for re-use elsewhere. I had no idea I could make `ref struct` extension methods haha!

wsugarman commented on FO-DICOM.Core/Serialization/JsonDicomConverter.cs:
```diff
@@ -15,7 +15,7 @@
 {
 
-
+    [Obsolete("Please use DicomJsonConverter instead.")]
```
Given this was just added, and to avoid a breaking change, I've marked this as `Obsolete`

mrbean-bremen reviewed (approved):
Thanks, as far as I can see, this all makes sense!

wsugarman commented:
@mrbean-bremen - I see I don't have write access, so I think I'll need your help merging this change.

mrbean-bremen commented:
@wsugarman - the policy here is to give the collaborators some time to look at a PR, especially as I myself still quite new to fo-dicom. If nothing comes up in a couple of days, I will just merge it.

wsugarman commented:
Ah ok! Sounds good!

amoerie reviewed (approved):
LGTM
